python/tk_framework_desktopserver/shotgun/api_v2.py

```diff
@@ -0,0 +1,836 @@
+        Gets state-related data for the site. Exactly what data this is depends
+        on the "browser_integration" core hook's "get_site_state_data" method,
+        which returns a list of dicts passed to the Shotgun Python API's find
+        method as kwargs. The data returned by this method is cached based on
+        the WSS connection key provided to the API's constructor at instantiation
+        time. This means that this data is queried from Shotgun only once per
+        unique WSS connection.
+
+        :returns: A list of Shotgun entity dictionaries.
+        :rtype: list
+        """
+        if "site_state_data" not in self.WSS_KEY_CACHE[self._wss_key]:
+            self.WSS_KEY_CACHE[self._wss_key]["site_state_data"] = []
+
+            requested_data_specs = self._engine.sgtk.execute_core_hook_method(
+                "browser_integration",
+                "get_site_state_data",
+            )
+
+            for spec in requested_data_specs:
+                entities = self._engine.shotgun.find(**spec)
+                self.WSS_KEY_CACHE[self._wss_key]["site_state_data"].extend(entities)
+        else:
+            self.logger.debug("Cached site state data found for %s" % self._wss_key)
+
+        return self.WSS_KEY_CACHE[self._wss_key]["site_state_data"]
+
+    def _get_software_entities(self):
+        """
+        Gets all Software entities from the Shotgun client site. Included are
+        all existing fields. This data is cached per WSS connection key that
+        was provided to the API's constructor at instantiation time. This means
+        that this data is queried from SHotgun only once per unique WSS
+        connection.
+
+        :returns: A list of Software entity dictionaries.
+        :rtype: list
+        """
+        if "software_entities" not in self.WSS_KEY_CACHE[self._wss_key]:
+            self.logger.debug(
+                "Software entities have not been cached for this connection, querying..."
+            )
+            self.WSS_KEY_CACHE[self._wss_key]["software_entities"] = self._engine.shotgun.find(
+                "Software",
+                [],
+                fields=self._engine.shotgun.schema_field_read("Software").keys(),
+            )
+        else:
+            self.logger.debug("Cached software entities found for %s" % self._wss_key)
+
+        return self.WSS_KEY_CACHE[self._wss_key]["software_entities"]
+
+    def _get_subprocess_kwargs(self):
+        """
+        Builds a list of kwargs to be passed to subprocesses.
+
+        :returns: A dict of kwargs.
+        :rtype: dict
+        """
+        kwargs = dict()
+
+        # If we're on Windows, we'll want to stop the cmd.exe window
+        # from flashing on/off when our subprocesses are run.
+        if sys.platform == "win32":
+            si = subprocess.STARTUPINFO()
+            si.dwFlags |= subprocess.STARTF_USESHOWWINDOW
+            kwargs["startupinfo"] = si
+
+        return kwargs
+
+    def _init_db(self):
```

Comment:
should this be in the caching thread, since this is the place where the db is written to? that keeps a clean writer/reader relationship between the caching code (this code) and the script that does the caching...